[WebGPU] Simplify MTLSamplerState cach

mwyrzykowski opened the pull request:
#### 6af2ed441c4ac89933830843181d5f2b332ca597
<pre>
[WebGPU] Simplify MTLSamplerState cach
<a href="https://bugs.webkit.org/show_bug.cgi?id=287580">https://bugs.webkit.org/show_bug.cgi?id=287580</a>
<a href="https://rdar.apple.com/144703694">rdar://144703694</a>

Reviewed by Cameron McCormack.

Simplify process side MTLSampler cache to do the following:

* remove custom type, instead use base64 String which is effectively the same
* limit number of process wide samplers to 2048 to avoid using too much
  memory in the GPU process on newer iPhones which have much larger limits.
  Limit is tracked by maxArgumentBufferSamplerCount so we take the smaller of
  the two: 2048 or maxArgumentBufferSamplerCount
* evict null entries from the cache when we reach half the maximum, to avoid unncessarily
  evicting MTLSamplerState instances which are in use.
* use HashMap with weak references to the id&lt;MTLSamplerState&gt; so we know
  when the ref count on the objects has reached zero

* LayoutTests/fast/webgpu/nocrash/fuzz-287580-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-287580.html: Added.
Add regresion test.

* Source/WebGPU/WebGPU/Sampler.h:
(WebGPU::Sampler::create):
* Source/WebGPU/WebGPU/Sampler.mm:
(WebGPU::computeDescriptorHash):
(WebGPU::Device::createSampler):
(WebGPU::Sampler::Sampler):
(WebGPU::Sampler::samplerState const):
(-[SamplerIdentifier initWithFirst:second:]): Deleted.
(-[SamplerIdentifier copyWithZone:]): Deleted.
(WebGPU::Sampler::~Sampler): Deleted.
Also switch from Objective-C containers to C++ containers.

* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::executePreDispatchCommands):
* Source/WebGPU/WebGPU/RenderBundle.h:
(WebGPU::RenderBundle::create):
* Source/WebGPU/WebGPU/RenderBundle.mm:
(WebGPU::RenderBundle::RenderBundle):
(WebGPU::RenderBundle::rebindSamplersIfNeeded const):
* Source/WebGPU/WebGPU/RenderBundleEncoder.h:
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::finish):
(WebGPU::RenderBundleEncoder::setBindGroup):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::executePreDrawCommands):
(WebGPU::RenderPassEncoder::executeBundles):
Ensure samplers are rebound prior to command buffer commit
call as opposed to encoding time.

Also, thank you for catching Darryl, RenderBundles failed to
rebind samplers which they internally used.

Canonical link: <a href="https://commits.webkit.org/290438@main">https://commits.webkit.org/290438@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64a23231fb22c62de7335f4c7b0fc3199c38d61c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90020 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9549 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44921 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95020 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40793 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9936 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17835 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69297 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26904 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93021 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7598 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81662 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49658 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7326 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39927 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77661 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37089 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96846 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17208 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12621 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78299 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17464 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77486 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77511 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19139 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21953 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20539 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10402 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17218 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22543 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16959 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20411 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18742 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam reviewed (approved):
r=me but let's mention in the commit message the ultimate aim here (to avoid creating too many simultaneous MTLSamplerState objects on the one MTLDevice) by making our cache hold weak references, and why we're limiting the cache size to 2048.

mwyrzykowski committed:
[WebGPU] Simplify MTLSamplerState cach
https://bugs.webkit.org/show_bug.cgi?id=287580
rdar://144703694

Reviewed by Cameron McCormack.

Simplify process side MTLSampler cache to do the following:

* remove custom type, instead use base64 String which is effectively the same
* limit number of process wide samplers to 2048 to avoid using too much
  memory in the GPU process on newer iPhones which have much larger limits.
  Limit is tracked by maxArgumentBufferSamplerCount so we take the smaller of
  the two: 2048 or maxArgumentBufferSamplerCount
* evict null entries from the cache when we reach half the maximum, to avoid unncessarily
  evicting MTLSamplerState instances which are in use.
* use HashMap with weak references to the id<MTLSamplerState> so we know
  when the ref count on the objects has reached zero

* LayoutTests/fast/webgpu/nocrash/fuzz-287580-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-287580.html: Added.
Add regresion test.

* Source/WebGPU/WebGPU/Sampler.h:
(WebGPU::Sampler::create):
* Source/WebGPU/WebGPU/Sampler.mm:
(WebGPU::computeDescriptorHash):
(WebGPU::Device::createSampler):
(WebGPU::Sampler::Sampler):
(WebGPU::Sampler::samplerState const):
(-[SamplerIdentifier initWithFirst:second:]): Deleted.
(-[SamplerIdentifier copyWithZone:]): Deleted.
(WebGPU::Sampler::~Sampler): Deleted.
Also switch from Objective-C containers to C++ containers.

* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::executePreDispatchCommands):
* Source/WebGPU/WebGPU/RenderBundle.h:
(WebGPU::RenderBundle::create):
* Source/WebGPU/WebGPU/RenderBundle.mm:
(WebGPU::RenderBundle::RenderBundle):
(WebGPU::RenderBundle::rebindSamplersIfNeeded const):
* Source/WebGPU/WebGPU/RenderBundleEncoder.h:
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::finish):
(WebGPU::RenderBundleEncoder::setBindGroup):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::executePreDrawCommands):
(WebGPU::RenderPassEncoder::executeBundles):
Ensure samplers are rebound prior to command buffer commit
call as opposed to encoding time.

Also, thank you for catching Darryl, RenderBundles failed to
rebind samplers which they internally used.

Canonical link: https://commits.webkit.org/290438@main
